File: GreekReader/app/src/main/java/com/ericmschmidt/greekreader/datamodel/Book.java
package com.ericmschmidt.greekreader.datamodel;

import java.util.ArrayList;

/**
 * Represents a sub-division within a larger text.
 */
public class Book {

    private ArrayList<String> _lines;
    private int _id;

    /**
     * Creates an instance of the Bok class
     */
    public Book(int id) {
        this._lines = new ArrayList<String>();
        this._id = id;
    }

    /**
     * Adds a line to the book.
     * @param line the line to add.
     */
    public void addLines(String line) {
        this._lines.add(line);
    }

    /**
     * Gets the ID of this book within the work.
     * @return an int specifying this book's position.
     */
    public int getId() {
        return this._id;
    }

    /**
     * Gets a line from this book.
     * @param position the position of the line from the beginning of the book.
     * @return the line.
     */
    public String getLine(int position) {

        String line = null;
        if (position >= 0
                && position < this._lines.size()) {
            line = this._lines.get(position);
        } else if (position > this._lines.size()) {

        } else {
            line = this._lines.get(0); // Default is to return the first line of the book.
        }
        return line;
    }

    /**
     * Gets the number of lines in this book.
     * @return an int of the number of lines.
     */
    public int getLineCount() {
        return this._lines.size();
    }
}

File: LatinReader/core/src/main/java/com/ericmschmidt/classicsreader/datamodel/Book.kt
package com.ericmschmidt.classicsreader.datamodel

/**
 * Represents a sub-division within a larger text.
 * @author Eric Schmidt
 * @author <a href="https://telpirion.com">...</a>
 * @version 2019-11-17
 * @since 1.0
 */
class Book(val id: Int) {

    private val _lines = mutableListOf<String>()

    /**
     * Adds a line to the book.
     * @param line the line to add.
     */
    fun addLines(line: String) {
        _lines.add(line)
    }

    /**
     * Gets a line from this book.
     * @param position the position of the line from the beginning of the book.
     * @return the line.
     */
    fun getLine(position: Int): String? {
        return if (position >= 0 && position < _lines.size) {
            _lines[position]
        } else if (position >= _lines.size) {
            null
        } else {
            _lines.firstOrNull() // Default is to return the first line of the book.
        }
    }

    /**
     * Get a sequence of lines from the book.
     * @param position the position to start the text from.
     * @param offset the number of lines to get.
     * @return String the text from the position indicated.
     */
    fun getLines(position: Int, offset: Int): String {
        if (position < 0) {
            return getLines(0, offset)
        }

        val toIndex = (position + offset).coerceAtMost(_lines.size)
        
        if (position > toIndex) {
            return ""
        }

        val subList = _lines.subList(position, toIndex)
        return if (subList.isEmpty()) {
            ""
        } else {
            subList.joinToString("\n", postfix = "\n")
        }
    }

    /**
     * Gets the number of lines in this book.
     * @return an int of the number of lines.
     */
    fun getLineCount(): Int {
        return _lines.size
    }
}
